### PR TITLE
Change `examples/virtio` to use new Makefile snippet system

### PR DESCRIPTION
**Pull request by alexandermbrown**
Similar to https://github.com/au-ts/libvmm/pull/77 but for `examples/virtio`.

### CODE REVIEW

**Comment by alexandermbrown** [comment]
@wom-bat would you be able to have a look at this PR as well?

File: examples/virtio/Makefile [hunk 1]
```diff
@@ -1,384 +1,64 @@
...
-CLIENT_VM_2_DTB := $(BUILD_DIR)/client_vm_2.dtb
-
-SDDF_BLK_COMPONENTS := $(SDDF_DIR)/blk/components
-SDDF_BLK_UTIL := $(SDDF_DIR)/blk/util
-
-ifeq ($(strip $(BOARD)), odroidc4)
-	export UART_DRIVER_DIR := meson
-else ifeq ($(strip $(BOARD)), qemu_arm_virt)
-	export UART_DRIVER_DIR := arm
+export override MICROKIT_SDK:=$(abspath $(MICROKIT_SDK))
```
Comment (wom-bat) [inline examples/virtio/Makefile:13]: Move into virtio.mk

File: examples/virtio/Makefile [hunk 2]
```diff
@@ -1,384 +1,64 @@
...
 
-$(BUILD_DIR)/postmark: $(LINUX_BLK_DIR)/postmark.c
-	$(CC_USERLEVEL) $(CFLAGS_LINUX) $< -o $@
+export BUILD_DIR:=$(abspath $(BUILD_DIR))
+export MICROKIT_SDK:=$(abspath $(MICROKIT_SDK))
+export VIRTIO_EXAMPLE:=$(abspath .)
 
-$(BUILD_DIR)/blk_client_init: $(LINUX_BLK_DIR)/blk_client_init
-	cp $< $@
+export TARGET := aarch64-none-elf
```
Comment (wom-bat) [inline examples/virtio/Makefile:40]: Move into virtio.mk

File: examples/virtio/board/odroidc4/virtio.system [hunk 3]
```diff
@@ -39,7 +39,7 @@
     </protection_domain>
 
     <protection_domain name="CLIENT_VMM-2" priority="100">
-        <program_image path="client_vmm_2.elf" />
+        <program_image path="client_vmm.elf" />
```
Comment (wom-bat) [inline examples/virtio/board/odroidc4/virtio.system:42]: The removal of separate client linux and roofs seems an unrelated change.

File: examples/virtio/Makefile [hunk 4]
```diff
@@ -1,384 +1,64 @@
...
+export CC := clang
+export CC_USERLEVEL := zig cc
```
Comment (wom-bat) [inline examples/virtio/Makefile:42]: Don't assume the existence of zig

**Review by wom-bat — CHANGES_REQUESTED** [review]

File: examples/virtio/Makefile (hunk 1 above)
Comment (alexandermbrown) [reply examples/virtio/Makefile:13]: Offline discussion: going to put this in another PR.

File: examples/virtio/Makefile (hunk 2 above)
Comment (alexandermbrown) [reply examples/virtio/Makefile:40]: Offline discussion: going to put this in another PR.

File: examples/virtio/Makefile (hunk 4 above)
Comment (alexandermbrown) [reply examples/virtio/Makefile:42]: See issue https://github.com/au-ts/libvmm/issues/57 - this is required at the moment

File: examples/virtio/board/odroidc4/virtio.system (hunk 3 above)
Comment (alexandermbrown) [reply examples/virtio/board/odroidc4/virtio.system:42]: I will move it into a separate PR. I still think client_vm_2 should be removed as it is identical to VM 1 and creates a lot of maintenance overhead

Comment (alexandermbrown) [reply examples/virtio/board/odroidc4/virtio.system:42]: Put it here https://github.com/au-ts/libvmm/pull/85

**Comment by alexandermbrown** [comment]
This is ready for review again but we need to review/merge https://github.com/au-ts/sddf/pull/152 before merging this one

File: examples/virtio/include/serial_config.h [hunk 5]
```diff
@@ -6,7 +6,7 @@
 #include <stdint.h>
 
 /* Number of clients */
-#define SERIAL_NUM_CLIENTS 3
+// #define SERIAL_NUM_CLIENTS 3
```
Comment (Ivan-Velickovic) [inline examples/virtio/include/serial_config.h:9]: Why this is commented out now?

File: examples/virtio/virtio.mk [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+VMM_TOOLS := $(LIBVMM)/tools
+MICROKIT_TOOL ?= $(MICROKIT_SDK)/bin/microkit
+SDDF_INCLUDE := $(SDDF)/include/sddf
+UTIL := $(SDDF)/util
+
+UART_DRIVER := $(SDDF)/drivers/serial/$(UART_DRIVER)
+SERIAL_COMPONENTS := $(SDDF)/serial/components
+BLK_COMPONENTS := $(SDDF)/blk/components
+
+SERIAL_NUM_CLIENTS := -DSERIAL_NUM_CLIENTS=3
```
Comment (Ivan-Velickovic) [inline examples/virtio/virtio.mk:12]: Put this in the header to be consistent with sDDF.

File: examples/virtio/include/serial_config.h (hunk 5 above)
Comment (alexandermbrown) [reply examples/virtio/include/serial_config.h:9]: Oh my bad that was before we made `NUM_CLIENTS` consistent

File: examples/virtio/virtio.mk (hunk 6 above)
Comment (alexandermbrown) [reply examples/virtio/virtio.mk]: Done

File: examples/virtio/include/serial_config.h (hunk 5 above)
Comment (alexandermbrown) [reply examples/virtio/include/serial_config.h]: Done

File: examples/virtio/Makefile (hunk 4 above)
Comment (alexandermbrown) [reply examples/virtio/Makefile:42]: Offline: going to do this in another PR since I'm only here for another 3 weeks

File: tools/linux/blk/blk_init.mk [hunk 7]
```diff
@@ -0,0 +1,21 @@
+#
+# Copyright 2024, UNSW
+#
+# SPDX-License-Identifier: BSD-2-Clause
+#
+
+ifeq ($(strip $(LIBVMM)),)
+$(error LIBVMM must be specified)
+endif
```
Comment (Ivan-Velickovic) [inline tools/linux/blk/blk_init.mk:9]: Why not do `LIBVMM ?= ../../`?

File: tools/linux/uio/uio.mk [hunk 8]
```diff
@@ -0,0 +1,43 @@
+#
+# Copyright 2024, UNSW
+#
+# SPDX-License-Identifier: BSD-2-Clause
+#
+# This Makefile snippet builds block UIO driver
+#
+
+ifeq ($(strip $(VMM_TOOLS)),)
```
Comment (Ivan-Velickovic) [inline tools/linux/uio/uio.mk:9]:
```suggestion
ifeq ($(strip $(LIBVMM_TOOLS)),)
```

Comment (Ivan-Velickovic) [reply tools/linux/uio/uio.mk:9]: Also why can't it be `LIBVMM_TOOLS ?= ../../`

Comment (alexandermbrown) [reply tools/linux/uio/uio.mk:9]: You can't use relative paths in snippets as the makefiles get copied and executed from an arbitrary build directory

File: tools/linux/blk/blk_init.mk (hunk 7 above)
Comment (alexandermbrown) [reply tools/linux/blk/blk_init.mk:9]: https://github.com/au-ts/libvmm/pull/79#discussion_r1685920469

File: tools/linux/uio/uio.mk (hunk 8 above)
Comment (alexandermbrown) [reply tools/linux/uio/uio.mk:9]: Will rename variable though

Comment (alexandermbrown) [reply tools/linux/uio/uio.mk]: Renamed

File: tools/linux/blk/blk_init.mk (hunk 7 above)
Comment (wom-bat) [reply tools/linux/blk/blk_init.mk:9]: You could do:
 ```
 LINUX_BLK_DIR := $(abs $(dir $(last ${MAKEFILES_LIST}))
 LIBVMM ?= $(realpath ${LINUX_BLK_DIR)}/../../)
 ```

File: tools/linux/blk/blk_init.mk [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+ifeq ($(strip $(MICROKIT_BOARD)),)
+$(error MICROKIT_BOARD must be specified)
+endif
+
+blk_client_init: $(LIBVMM)/tools/linux/blk/blk_client_init
```
Comment (wom-bat) [inline tools/linux/blk/blk_init.mk:14]: Might be better to put all the files you want to dump into the root image into a subdir, one for each root image.
Maybe note and do that later

File: tools/linux/uio/uio.mk [hunk 10]
```diff
@@ -0,0 +1,43 @@
+#
+# Copyright 2024, UNSW
+#
+# SPDX-License-Identifier: BSD-2-Clause
+#
+# This Makefile snippet builds block UIO driver
+#
```
Comment (wom-bat) [inline tools/linux/uio/uio.mk:7]: Include the name of the thing(s) it builds, as somewhere they need to be added to a dependency in the top-level makefiles.

Comment (wom-bat) [reply tools/linux/uio/uio.mk:7]: This snippet actually builds libuio.a --- a library that will be linked against the actual drivers.

**Review by wom-bat — CHANGES_REQUESTED** [review]

File: tools/linux/blk/blk_init.mk (hunk 9 above)
Comment (alexandermbrown) [reply tools/linux/blk/blk_init.mk:14]: I'm not sure how we'd do this in `blk_init.mk`. Scripts like `blk_client_init` would need to be copied into more than one rootfs. `blk_init.mk` does not know the layout of our system.

File: tools/linux/uio/uio.mk (hunk 10 above)
Comment (alexandermbrown) [reply tools/linux/uio/uio.mk:7]: Done https://github.com/au-ts/libvmm/pull/79/commits/d648c3778208ea2e4e3f443439db11041b7cf325

File: tools/linux/blk/blk_init.mk (hunk 7 above)
Comment (alexandermbrown) [reply tools/linux/blk/blk_init.mk]: Done https://github.com/au-ts/libvmm/pull/79/commits/282cad92d36ff3fe73ecd9c747e7ae1a08ab7717

**Comment by Ivan-Velickovic** [comment]
I think all the feedback has been addressed now, can you squash the commits into one and then we can merge.

File: tools/linux/blk/blk_init.mk (hunk 9 above)
Comment (alexandermbrown) [reply tools/linux/blk/blk_init.mk:16]: Offline discussion: going to do this later, maybe as a part of the Lions OS Makefile generation.
